### ISSUE TITLE
Numerical input and output does not satisfy the main invariant of section 6.2.7 of the R7RS

### ISSUE THREAD

**Issue by mnieper** [ashinn/chibi-scheme#433]
For example,

```scheme
(let ((number 1.448997445238699))
    (eqv? number
          (string->number (number->string number))))
```
evaluates to `#f`.

(I haven't checked the underlying conversion function, but it seems that `number->string` is not accurate enough. It yields `1.4489974452387`, while other Schemes where the above test does not fail (I checked with Chez Scheme and with Scheme48) yield the string `1.448997445238699`.)

**Comment by mnieper** [comment]
The following algorithm by Aubrey Jaffer is easy to implement: http://people.csail.mit.edu/jaffer/III/EZFPRW. In the example above, it gives the correct result.

**Comment by ashinn** [comment]
You're assuming the loss of precision is in write, while it may be in read or in both.  write simply falls back to sprintf, which has no such guarantees.

In fmt (http://synthcode.com/scheme/fmt/) I implemented Dybvig's algorithm, and it gets this right.  The simplification in (chibi show) falls back to number->string, so fails for Chibi here.

I never really intended the core reader/writer for Chibi to be accurate in this respect, valuing very fast startup time above all else, but I might consider updating (scheme write) with a slower but more accurate version.

**Comment by mnieper** [comment]
Of course, the loss of precision may also be in read. It was just a guess from a few numerical experiments.

For startup times and reading code, the speed of reading/writing inexact numbers does not matter much, so it won't be a loss if the core reader/writer would become accurate in this respect. (It would matter much more if the program to load contained lots and lots of inexact constants, but for such a program, it would probably also be very important that the numbers are read in accurately.)

**Comment by mnieper** [comment]
16 decimal places of precision are not enough to represent every double. The following example fails:

```scheme
(let ((number 0.14285714285714282))
    (eqv? number
          (string->number (number->string number))))
```

**Comment by ashinn** [comment]
That was actually me being lazy and getting you to find a counter-example :)

Since C has no requirement of the minimal representation, using 17 decimal places results in annoyances like 0.1 printing as 0.10000000000000001.  I can probably fix this with some post-processing.

**Comment by mnieper** [comment]
Alex Shinn <notifications@github.com> schrieb am Mo., 21. Aug. 2017 um
16:45 Uhr:

> That was actually me being lazy and getting you to find a counter-example
> :)
>
That's called efficient division of labor! :-)

> Since C has no requirement of the minimal representation, using 17 decimal
> places results in annoyances like 0.1 printing as 0.10000000000000001. I
> can probably fix this with some post-processing.
>
You can probably use the same idea used by Aubrey Jaffer. Round to 16
places, check whether this reads back as the correct number. If not, round
to 17 places.

Marc

**Comment by ashinn** [comment]
I've pushed a fix in d9a40fbc615e31f6e4663177c6f7ed1e1c0406de that will fall back to 17 digits if needed.  I've also improved read precision in the most obvious way by computing with long doubles.

Chibi still doesn't guarantee the shortest possible representation, though I may implement this at some point for (scheme write).
